### PR TITLE
hyperparameter tuning 

### PR DESCRIPTION
**Pull request by prithvikannan**
## Related Issues/PRs

building on https://github.com/mlflow/mlflow/pull/6680, blocked by https://github.com/mlflow/mlp-regression-template/pull/75

## What changes are proposed in this pull request?

Filling in the `objective()` function in HP tuning and handle hardcoded parameters. Some small refactors to reduce code repetition

## How is this patch tested?

Manual testing using mlp-regression-template repo and a source mlflow installation.

- [x] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/pipelines/steps/train.py [hunk 1]
```diff
@@ -124,135 +124,181 @@ def _run(self, output_directory):
...
+                def objective(hyperparameter_args):
+                    # log as a child run
+                    with mlflow.start_run(tags=tags, nested=True):
+                        # create unfitted estimator from yaml
+                        estimator = estimator_fn(
+                            dict(estimator_hardcoded_params, **hyperparameter_args)
+                        )
+                        sample_fraction = (
+                            float(tuning_params["sample_fraction"])
+                            if "sample_fraction" in tuning_params
```
Comment (dbczumar) [inline mlflow/pipelines/steps/train.py:152]: Do we validate that sample fraction is > 0 and <= 1?

File: mlflow/pipelines/steps/train.py [hunk 2]
```diff
@@ -124,135 +124,181 @@ def _run(self, output_directory):
...
+                            else 1.0
+                        )
+                        X_train_sampled = X_train.sample(frac=sample_fraction, random_state=42)
+                        y_train_sampled = y_train.sample(frac=sample_fraction, random_state=42)
+                        # fit estimator to training
+                        estimator.fit(X_train_sampled, y_train_sampled)
+                        if hasattr(estimator, "best_score_P"):
```
Comment (dbczumar) [inline mlflow/pipelines/steps/train.py:159]:
```suggestion
                        if hasattr(estimator, "best_score_"):
```

File: mlflow/pipelines/steps/train.py [hunk 3]
```diff
@@ -124,135 +124,181 @@ def _run(self, output_directory):
...
+                            mlflow.log_metric("best_cv_score", estimator.best_score_)
+                        if hasattr(estimator, "best_params_"):
+                            mlflow.log_params(estimator.best_params_)
```
Comment (dbczumar) [inline mlflow/pipelines/steps/train.py:162]: Is it ever possible for these best params / best score to be invalid (e.g. `None`)? Shall we validate them?

Comment (dbczumar) [reply mlflow/pipelines/steps/train.py:162]: Ah, I see. This is duplicated from below. Can we factor it out?

File: mlflow/pipelines/steps/train.py [hunk 4]
```diff
@@ -124,135 +124,181 @@ def _run(self, output_directory):
...
+                                self.pipeline_root,
+                                self.evaluation_metrics.values(),
+                            ),
+                            evaluator_config={
+                                "log_model_explainability": False,
+                            },
+                        )
+
+                        # return +/- metric
+                        return eval_result.metrics[self.primary_metric]
```
Comment (dbczumar) [inline mlflow/pipelines/steps/train.py:192]: Hyperopt needs to know whether the objective is to minimize or maximize this metric based on the "greater_is_better" value associated with the metric. Can we incorporate that into this PR? This needs to be supported for predefined metrics and custom metrics.

File: mlflow/pipelines/steps/train.py [hunk 5]
```diff
@@ -124,135 +124,181 @@ def _run(self, output_directory):
...
-                    param_details_to_pass.pop("distribution")
-                    search_space[param_name] = hp_tuning_fn(param_name, **param_details_to_pass)
-                else:
+        mlflow.autolog(log_models=False)
+        with mlflow.start_run(tags=tags) as run:
+            estimator_hardcoded_params = self.step_config["estimator_params"]
+            if self.step_config["tuning_enabled"]:
+                # gate all HP tuning code within this condition
+                tuning_params = self.step_config["tuning"]
+                # import hyperopt or throw error
```
Comment (dbczumar) [inline mlflow/pipelines/steps/train.py:133]: Nit: can we remove this comment? Self-explanatory from the following lines

File: mlflow/pipelines/steps/train.py [hunk 6]
```diff
@@ -124,135 +124,181 @@ def _run(self, output_directory):
...
-                    hp_tuning_fn = getattr(hp, param_details["distribution"])
-                    param_details_to_pass = param_details.copy()
-                    param_details_to_pass.pop("distribution")
-                    search_space[param_name] = hp_tuning_fn(param_name, **param_details_to_pass)
-                else:
+        mlflow.autolog(log_models=False)
+        with mlflow.start_run(tags=tags) as run:
+            estimator_hardcoded_params = self.step_config["estimator_params"]
+            if self.step_config["tuning_enabled"]:
+                # gate all HP tuning code within this condition
```
Comment (dbczumar) [inline mlflow/pipelines/steps/train.py:131]: Nit: can we remove this comment? Not particularly necessary

File: mlflow/pipelines/steps/train.py [hunk 7]
```diff
@@ -124,135 +124,181 @@ def _run(self, output_directory):
...
-                logged_estimator = mlflow.sklearn.log_model(
-                    estimator,
-                    f"{self.name}/estimator",
-                    signature=estimator_schema,
-                    code_paths=code_paths,
+                # wrap training in objective fn
+                def objective(hyperparameter_args):
+                    # log as a child run
+                    with mlflow.start_run(tags=tags, nested=True):
+                        # create unfitted estimator from yaml
```
Comment (dbczumar) [inline mlflow/pipelines/steps/train.py:146]: Can we remove these comments unless the code that follows them is particularly hard to read / unintuitive?

File: mlflow/pipelines/steps/train.py [hunk 8]
```diff
@@ -124,135 +124,181 @@ def _run(self, output_directory):
...
+                        estimator = estimator_fn(
+                            dict(estimator_hardcoded_params, **hyperparameter_args)
+                        )
+                        sample_fraction = (
+                            float(tuning_params["sample_fraction"])
+                            if "sample_fraction" in tuning_params
+                            else 1.0
+                        )
+                        X_train_sampled = X_train.sample(frac=sample_fraction, random_state=42)
+                        y_train_sampled = y_train.sample(frac=sample_fraction, random_state=42)
```
Comment (dbczumar) [inline mlflow/pipelines/steps/train.py:156]: Is this sampling stable such that it will always pick the same records given the same dataset? If not, we should investigate using the sampling logic from the MLP split step: https://github.com/mlflow/mlflow/blob/eb693368c00595b6d41f42759c6d2243430534f7/mlflow/pipelines/steps/split.py#L194-L195

File: mlflow/pipelines/steps/train.py [hunk 9]
```diff
@@ -124,135 +124,181 @@ def _run(self, output_directory):
...
-                elif "distribution" in param_details:
-                    hp_tuning_fn = getattr(hp, param_details["distribution"])
-                    param_details_to_pass = param_details.copy()
-                    param_details_to_pass.pop("distribution")
-                    search_space[param_name] = hp_tuning_fn(param_name, **param_details_to_pass)
-                else:
+        mlflow.autolog(log_models=False)
+        with mlflow.start_run(tags=tags) as run:
+            estimator_hardcoded_params = self.step_config["estimator_params"]
+            if self.step_config["tuning_enabled"]:
```
Comment (dbczumar) [inline mlflow/pipelines/steps/train.py:130]: As discussed offline, can we factor out the tuning logic into multiple helper methods that are invoked from within this branch and deduplicate some of the code that might be shared with the non-tuning branch?

File: mlflow/pipelines/steps/train.py (hunk 8 above)
Comment (prithvikannan) [reply mlflow/pipelines/steps/train.py:150]: @dbczumar looks like `pandas.DataFrame.sample` is stable
> If you pass it a RandomState it will use this (already initialized/seeded) RandomState to generate pseudo random numbers. This also allows you to get reproducible results by setting a fixed seed when initializing the RandomState and then passing this RandomState around.

source: https://stackoverflow.com/a/45213353

File: mlflow/pipelines/steps/train.py (hunk 4 above)
Comment (prithvikannan) [reply mlflow/pipelines/steps/train.py:173]: @dbczumar i'm not quite sure how to solve this. the only documentation i see about `greater_is_better` is in [sklearn.metrics.make_scorer](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html), which i don't think is useful here.

what i'm looking for is something like `is_metric_greater_is_better(metric_name)`, as in [catboost](https://github.com/catboost/catboost/issues/1890), but i don't think this is supported in sklearn.

File: mlflow/pipelines/steps/train.py (hunk 8 above)
Comment (dbczumar) [reply mlflow/pipelines/steps/train.py:150]: Yay!

File: mlflow/pipelines/steps/train.py (hunk 4 above)
Comment (dbczumar) [reply mlflow/pipelines/steps/train.py:173]: Ah, the user specifies `greater_is_better` for custom metrics here: https://github.com/mlflow/mlp-regression-template/blob/74284ee13d907b0b427fd76d144986ac333d520f/pipeline.yaml#L85. We just need to read this value.

For builtin metrics, `greater_is_better` is defined here: https://github.com/mlflow/mlflow/blob/7219771141c7140fd68f6ed3a5b649e5ad1dab13/mlflow/pipelines/utils/metrics.py#L40-L44

File: mlflow/pipelines/steps/train.py [hunk 10]
```diff
@@ -124,135 +125,149 @@ def _run(self, output_directory):
...
+
+                best_hp_params = fmin(
+                    objective,
+                    search_space,
+                    algo=tuning_algo,
+                    max_evals=max_trials,
+                    trials=hp_trials,
                 )
+                best_hp_estimator_loss = hp_trials.best_trial["result"]["loss"]
+                hardcoded_estimator_loss = objective({})
```
Comment (dbczumar) [inline mlflow/pipelines/steps/train.py:189]: This won't work if the user-defined model returned by `estimator_fn` initializes model parameters using the args passed to `estimator_fn`. We should instead call `objective()` with the hardcoded parameter dict, rather than an empty dict.

File: mlflow/pipelines/steps/train.py (hunk 9 above)
Comment (dbczumar) [reply mlflow/pipelines/steps/train.py:131]: @prithvikannan Can we move the tuning logic into a separate method so that we can just call

```
if self.step_config["tuning_enabled"]:
    estimator = tune_and_get_best_estimator(...)
else:
    estimator = estimator_fn(estimator_hardcoded_params)
```

?

File: mlflow/pipelines/steps/train.py [hunk 11]
```diff
@@ -548,3 +530,110 @@ def environment(self):
...
+                )
+
+                # return +/- metric
+                sign = -1 if self.evaluation_metrics_greater_is_better[self.primary_metric] else 1
+                return sign * eval_result.metrics[self.primary_metric]
+
+        search_space = self._construct_search_space_from_yaml(tuning_params["parameters"])
+        algo_type, algo_name = tuning_params["algorithm"].rsplit(".", 1)
+        tuning_algo = getattr(importlib.import_module(algo_type, "hyperopt"), algo_name)
+        max_trials = tuning_params["max_trials"]
```
Comment (dbczumar) [inline mlflow/pipelines/steps/train.py:582]: Can we add validation in `from_pipeline_config` to ensure that these `parameters`, `algorithm`, and `max_trials` fields are defined? For algorithm, I think we can default to hyperopt random search, so that field can be optional.

File: mlflow/pipelines/steps/train.py [hunk 12]
```diff
@@ -523,9 +485,29 @@ def from_pipeline_config(cls, pipeline_config, pipeline_root):
                         "'enabled' key whose value is either true or false.",
                         error_code=INVALID_PARAMETER_VALUE,
                     )
+
+                if "sample_fraction" in step_config["tuning"]:
+                    sample_fraction = float(step_config["tuning"]["sample_fraction"])
+                    if sample_fraction >= 0 and sample_fraction <= 1.0:
```
Comment (dbczumar) [inline mlflow/pipelines/steps/train.py:491]:
```suggestion
                    if sample_fraction > 0 and sample_fraction <= 1.0:
```

Comment (dbczumar) [reply mlflow/pipelines/steps/train.py:491]: Sample fraction of zero shouldn't be allowed

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @prithvikannan !

File: mlflow/pipelines/steps/train.py (hunk 11 above)
Comment (prithvikannan) [reply mlflow/pipelines/steps/train.py:582]: updated. already had `algorithm` set to hp random as default, but added validation for `parameters` and `max_trials`.
